semantic-core/src/Core/Parser.hs

```diff
@@ -1,29 +1,38 @@
-{-# LANGUAGE FlexibleContexts, TypeOperators #-}
+{-# LANGUAGE DeriveFunctor, DerivingStrategies, FlexibleContexts, GeneralizedNewtypeDeriving, TypeOperators #-}
 module Core.Parser
   ( core
   , lit
   , expr
   , record
   , comp
-  , lvalue
   ) where
 
 -- Consult @doc/grammar.md@ for an EBNF grammar.
 
 import           Control.Applicative
 import           Control.Effect.Carrier
+import           Control.Monad
 import           Core.Core ((:<-) (..), Core)
 import qualified Core.Core as Core
 import           Core.Name
 import qualified Data.Char as Char
 import           Data.Foldable (foldl')
 import           Data.String
+import           Text.Parser.LookAhead (LookAheadParsing)
 import qualified Text.Parser.Token as Token
 import qualified Text.Parser.Token.Highlight as Highlight
+import qualified Text.Parser.Token.Style as Style
 import           Text.Trifecta hiding (ident)
 
 -- * Identifier styles and derived parsers
 
+newtype CoreParser m a = CoreParser { runCoreParser :: m a }
+  deriving stock Functor
+  deriving newtype (Alternative, Applicative, CharParsing, DeltaParsing, Errable, LookAheadParsing, Monad, MonadPlus, Parsing)
+
+instance TokenParsing m => TokenParsing (CoreParser m) where
+  someSpace = Style.buildSomeSpaceParser (void (satisfy Char.isSpace)) Style.haskellCommentStyle
```

Comment:
Can we go with `#` for line comments & no block comments?

1. We don’t have editor support and no editor on earth is going to go with `--` knowing nothing else about the language.

2. Per Kmett, we might stand a chance of parsing this as a Dyck language (and therefore monoidally) if we disallow nesting block comments.

Reply:
If we want to go with `#` for block comments, we’ll have to rework `#unit{}` and `#record{}`. `//` would be an alternative that didn’t require such.

Reply:
We talked about this a bit and `#` will conflict with a bunch of the extant syntax so we’re not gonna sweat it for now.